Zsuk

ZeyaGun opened the pull request.

ZeyaGun committed:
no message

ZeyaGun committed:
no message

ZeyaGun committed:
refactor : TrackPlayer.cs

ZeyaGun committed:
rename : MovePlayer

Cinemachine 스크립트 PlayerMove와 중복되어 수정

ZeyaGun committed:
상호작용 관련 예시 스크립트 구현

ZeyaGun committed:
feat: StageController

스테이지에 관련된 정보를 저장, 관리하는 StageController를 임시로 구현하였음
수정 가능성 있음

ZeyaGun committed:
no message

ZeyaGun committed:
feat: Controller

디버깅용 컨트롤러 (간단한 테스트용)

ZeyaGun committed:
refactor : StageController.cs

Stage 클래스에 캐릭터를 시작할 위치 정보를 가진 Postiion 변수를 추가, 관련 프로퍼티 추가, 생성자 수정

ZeyaGun committed:
feat  : Controller.cs
단순 파일 위치 수정

ZeyaGun committed:
동일

ZeyaGun committed:
feat : dummy_ji_2

기존 배치를 바꾸고자 새로운 더미 씬 제작

ZeyaGun committed:
feat: ObjectInteractable

상호작용 가능한 오브젝트 클래스

ZeyaGun committed:
feat: InteractionPlayer.cs

현재 상호작용 가능한 오브젝트를 큐에 저장해서

상호작용시 사용하는 클래스

ZeyaGun committed:
feat: Controller

제작 시 편의를 위해 임의로 스테이지를 바꾸는 기능 구현

ZeyaGun committed:
add: Cinemachine

시네머신을 임포트 하였음

jsuk10 commented on Assets/Scripts/Controller.cs:
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Controller : MonoBehaviour
+{
+    void Start() 
+    {
+        StageController.instance.SetStage(0);
+    }
+    void Update()
+    {
+        InputProcess();
+    }
+
+    void InputProcess()
+    {
+        if(Input.GetKeyDown(KeyCode.Alpha0))
+        {
+            StageController.instance.SetStage(0);
+            Debug.Log(0);
+        }
+        else if(Input.GetKeyDown(KeyCode.Alpha1))
+        {
+            StageController.instance.SetStage(1);
+             Debug.Log(1);
+        }
+    }
+}
```
테스트용 코드는 올려주지 않는게 좋을것 같습니다.

ZeyaGun replied:
참고하겠습니다

ZeyaGun commented:
플레이어와 오브젝트의 상호작용 스크립트 + 편의상 스테이지 변경 등 기능을 작업하였음

플레이어와 오브젝트의 상호작용은 
플레이어와 오브젝트에 충돌체를 주어

충돌시 큐에 삽입하여 동시에 두개 이상의 오브젝트에 충돌하는 상태를 해결하였고, 
충돌 해제시 오브젝트가 큐에게 스스로를 반환 요청하여 더이상 상호작용 가능하지 않게 하였음

jsuk10 commented on Assets/Scripts/InteractionPlayer.cs:
```diff
@@ -0,0 +1,53 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class InteractionPlayer : MonoBehaviour
```
모든 오브젝트에 대한 이벤트를 관리 하는 거면 싱글톤으로 하는게 낫지 않을까요?

ZeyaGun replied:
수정할 예정입니다

ZeyaGun replied:
수정하였습니다

jsuk10 commented on Assets/Scripts/Object/ObjectInteractable.cs:
```diff
@@ -0,0 +1,64 @@
+    /// 삽입 여부를 참으로 전환하고 상호작용 큐에 삽입 
+    /// </summary>
+    /// <param name="other"></param>
+    private void OnTriggerStay(Collider other) 
```
onStay 보다 enter이 호출 횟수가 적을것 같은데 stay보다 enter로 하는게 좋지 않나요?
이렇게 하신 이유가 있나요??

ZeyaGun replied:
지속적으로 상호작용가능한 오브젝트의 경우를 고려하여 stay로 지속적으로 큐에 넣어주는 방식으로 구현하였습니다

jsuk10 commented on Assets/Scripts/Object/ObjectInteractable.cs:
```diff
@@ -0,0 +1,64 @@
+            if(isEnqueued == false)
+            {
+                isEnqueued = true;
+                var player = other.gameObject.GetComponent<InteractionPlayer>();
+                player.EnqueueObject(this);
```
getComponent가 비용이 큰걸로 아는데 stay에서 실행하는거면 싱글톤이 좋아 보입니다.

ZeyaGun replied:
위의 InteractionPlayer.cs를 수정하면 해결될 것 같습니다

ZeyaGun replied:
수정하였습니다

jsuk10 commented on Assets/Scripts/Object/ObjectInteractable.cs:
```diff
@@ -0,0 +1,64 @@
+    private void OnTriggerExit(Collider other) 
+    {
+        Debug.Log(gameObject.transform.parent.gameObject.name + " Exit");
+        if(other.tag == "Player")
+        {
+            var player = other.gameObject.GetComponent<InteractionPlayer>();
+            player.DequeueObject();
+            isEnqueued = false;
+        }
```
Queue가 좋지 않은 코드인것 같습니다.
예를들어 1번과 2번이 들어간 뒤에 2번의 범위밖으로 나가게 되어 dequeue하게 되면 2번이 아닌 1번이 빠저나가는 오류가 생길 수 있을 것 같습니다. 
queue말고 다른 자료구조를 사용해야 할 것 같습니다.
저는 개인적으로 String bool 딕셔너리를 추천합니다.

ZeyaGun replied:
조금 생각이 필요합니다 
기본적으로는 순서에 따라 접근하되 제거 시에만 탐색을 할 것 같아서
탐색 효율이 높고 추가속도가 느린 딕셔너리보다
추가 제거가 빠르고 탐색 효율은 떨어지는 List가 용이할 듯 합니다

ZeyaGun replied:
수정하였습니다

jsuk10 commented on Assets/Scripts/StageController.cs:
```diff
@@ -9,10 +9,11 @@
+        position = Position;
     }
     GameObject vCam;
     public GameObject VCam
```
Class나 enum은 꼭 필요한 경우가 아니면 전용 스크립트에 쓰면 좋을것 같습니다.
Script/DataSet에 Class, enum을 만들어 두었으니 차후에 옮겨 주시면 감사하겠습니다.

ZeyaGun replied:
차후 수정하겠습니다

ZeyaGun replied:
수정하였습니다

jsuk10 commented on Assets/Scripts/StageController.cs:
```diff
@@ -68,6 +80,8 @@ public void SetStage(int currentStageNum)
 
```
enum class에 stage를 기록했으니 enum을 사용하는건 어떨까요?

ZeyaGun replied:
차후 수정하겠습니다

jsuk10 commented on Assets/Scripts/Player/MovePlayer.cs:
```diff
@@ -0,0 +1,80 @@
+            moveDirection.Normalize();
+            return true;
+        }
+    }
```
moveDirection = new Vector3(Input.GetAxis("Vertical"), Input.GetAxis("Horizontal"), 0 );
을 한 뒤에 정규화 하면 위의 코드를 한번에 줄일수 있지 않을까요?

ZeyaGun replied:
코드가 의도한 바를 만족시키면서 성능을 향상시키는것으로 생각되어 수정하겠습니다

ZeyaGun replied:
의도한 대로 동작하지 않아서 차후에 나은 방법을 찾아보도록 하겠습니다

jsuk10 reviewed (commented):
리뷰 하였습니다.

ZeyaGun committed:
feedback : ObjectInteractable.cs

피트백에 따라 InteractionPlayer.cs를 싱글톤으로 만들고
해당 스크립트 내부의 GetConponent문 대신 싱글톤을 사용하여 인스턴스 호출

ZeyaGun committed:
feedback: InteractionPlayer.cs

미리 만들어진 Singleton Generic Class를 사용하여 싱글톤으로 바꿔주었음

ZeyaGun committed:
feedback: InteractionPlayer.cs

queue에서 list로 바꾸고 그로 인해 생긴 문제를 해결하였음

ZeyaGun committed:
feedback : ObjectInteractable

InteractionPlayer.cs가 queue에서 list로 바꾸면서 바뀐 함수명을 사용하도록 바꿔주었습니다

ZeyaGun committed:
Merge branch 'develop' into Zsuk

ZeyaGun committed:
Merge branch 'Zsuk' of https://github.com/Leejeongseok0407/HOME into Zsuk

ZeyaGun committed:
chore : etc

Init 대소문자 수정

ZeyaGun committed:
chore : Stage.cs

StageController에 있는 Stage Class를
Class.cs로 이동

jsuk10 reviewed (dismissed):
고생하셨습니다.

ZeyaGun committed:
chore : etc
